### ISSUE TITLE
Add "ignore_changes" support

### ISSUE THREAD

**Issue by mickledur** [Azure/azure-service-operator#2938]
Hi, 
We have a use case where a Redis Cache is cretaed by ASO.
Once created we have an external support team that need to put some tags manually on the Redis Cache.
But since ASO compares current state with desired state each 30min, the tags get removed.
Support team can't give us the tags in advance since they don't know all of them in advance. Otherwise we could add them when deploying the Redis Cache.

As workaround, when the Redis Cache is deployed we put the annotation "serviceoperator.azure.com/reconcile-policy: skip". But this is not an efficient solution since it brings other issues.

What we want is to have a parameter at the Resource level which allows  to ignore changes affecting an attribute of the resource. Something like what is done with Terraform parameter " ignore_changes = [tags]".

Any idea on how to implement this?

Thanks in advance for your support

**Comment by matthchr** [comment]
Sorry for the slow response, I wanted to make sure that I understood exactly how `ignore_changes` was implemented for Terraform and it took me a while to get a chance to dig in and understand how exactly it works.

In the short term, can your support team make their updates via Kuberentes? If they change the resource in Kubernetes and it doesn't have the `skip` annotation set, the tags will correctly get pushed to Azure immediately.

In the longer term, we can talk about implementing this as a feature. It clearly gets pretty heavy use in Terraform and I see the use-case for it here as well, although I do think it's somewhat more awkward in the Kubernetes context than it is in the Terraform context.

The way that Terraform does this is that they start with [the existing resource in Azure](https://github.com/hashicorp/terraform-provider-azurerm/blob/e242e2c506f54f4f26e74e23d02453c5392bdf47/internal/services/containers/kubernetes_cluster_resource.go#L1595), and then [merge in changes from the user model](https://github.com/hashicorp/terraform-provider-azurerm/blob/e242e2c506f54f4f26e74e23d02453c5392bdf47/internal/services/containers/kubernetes_cluster_resource.go#L1975) over the top of that if `ignore_changes` is not set for that field.

What this basically boils down to is:
1. HTTP GET
2. Modify GET payload with select contents from Terraform files (honoring `ignore_changes`)
3. HTTP PUT

This is in contrast to what ASO does today, which is:
1. Build HTTP PUT payload based on CR state
2. HTTP PUT

So we would need to universally HTTP GET first, and then have some collection of fields (JSONPath expressions, maybe?) the user could specify which we would then merge into the HTTP PUT payload prior to actually sending it. That would make ASO something like:

1. HTTP GET
2. Build HTTP PUT payload based on CR state
3. Update HTTP PUT payload based on `ignore_changes` (getting value from HTTP GET instead).
4. HTTP PUT

It's doable but quite a bit of work. I do think there's probably value in doing it but I'd like to see some more community feedback/interest in the topic before we commit to the work.

**Comment by mickledur** [comment]
Thank you @matthchr for your answer. Very clear.
We will go with the workaround to satisfy our need.
I will keep watchig the RoadMap of ASO to see when you will implement this feature. 
Hope the community will have interest in this topic.
Thanks you again for your time.

**Comment by matthchr** [comment]
We're not 100% convinced we can/should implement this, but not closing this as we want to see how much user need there is.

**Comment by matthchr** [comment]
No change from above.

**Comment by matthchr** [comment]
Still gathering community feedback on this one.

**Comment by matthchr** [comment]
No change from above (but we haven't forgotten about this)

**Comment by matthchr** [comment]
We're discussing if we can do this generically or if support for a more limited `ignore_changes` that only worked for tags, which are a common point of coordination and often the reason that folks want `ignore_changes`.

things to consider:
- eTag support -- not every Azure resource supports eTags, because it's not required by ARM. Without eTag, it's not really _safe_ to GET + modify + PUT, even if it might work most of the time.
- We have another item which we've already got an accepted design review for (#4962) which is proposing a solution to issue #2811. We may want to implement that solution first, as it also will come with GET infrastructure which would likewise be required for this ignore_changes support.

We're curious for input from the community on anybody waiting for this feature, would tags-only ignore_changes solve your problems?

**Comment by antonioquintavalle1A** [comment]
I believe this is the general intent of the `skip` annotation; it guides users understanding that resources are effectively `managed` outside of ASO, even if it's just for its own labels.
As long as the resource is created only once and it does not require future updates `ignore_changes` here would be useless.

Another consideration to taken in account is how people are supposed to manage their ASO resources.
In a gitops approach, for instance, I believe the right way would be to have teams contributing to the same repository, hence with simple code changes resources can be updated correctly.
@MZODEH-se, in the scenario you're describing would make sense that the `platform` team creates a Redis Cache without labels through an ASO object and then the external support team commits their labels on the repo so that they can get aligned in the next reconciliation loop?

Switching to this paradigm will give you much more control of the objects; external team won't need to interact with K8S cluster and changes to ASO resources will be tracked and reviewed with git commits/PRs.

Doing so, ASO can stick to the current paradigm enfoced by the `manage` / `skip` annotations, you'll be using the standard `manage` one but the external team will make a commit in GIT instead of patching the azure object directly.

How does it sound like?

**Comment by matthchr** [comment]
Yes we generally agree @antonioquintavalle1A that in an ideal gitops world, `skip` is enough. That's part of the reason why we haven't implemented `ignore_changes` -- the other part being that actually implementing it is a bit tricky to get right in the context of ASO.

It's something we're interested in more community input on as well.
